libs/core/langchain_core/indexing/schema_02.py

```diff
@@ -0,0 +1,265 @@
+"""Proposal for a new schema for the indexing and retrieval interfaces.
+
+This proposal has the following goals:
+
+1. Have a well-defined READ/WRITE API for indexing.
+    - Use upsert instead of add
+    - Support upsert, delete and read by ID
+2. Support indexing of binary data.
+3. Have a well-defined query API on retrieval
+    - Define filters, limit, offset, sort
+    - Support similarity search
+    - Enforcement can be done by standard unit-tests against the implementation.
+    - Implementations can be required to either self-describe supported features
+      or else just raise a NotImplementedError for unsupported features, so
+      no silent failures occur.
+    - We may need to throw in an extra field for disambiguating between
+      a query executing using the `raw` interface vs. via a standardized langchain
+      interface.
+
+The query API should allow us to support adhoc capabilities like:
+    - mmr search
+    - setting min score threshold etc
+
+-----
+
+Discussion points:
+
+* Force all content to have metadata OR only ID?
+    - If it's only ID, then we don't support the ability to filter by metadata in the query API.
+    Abstraction is more generic, and we can sub-class to add an interface that supports metadata operations.
+* Base Indexing abstraction forced to support metadata?
+* Support get_with_extras for indexers or vectorstores? (See below)
+
+*. Two options for how to deal with the retriever interface:
+    1. Have a standalone interface, and a vectorstore implements both indexer and retriever interfaces.
+    2. The retriever interface is defined in the vectorstore abstraction itself (which inherits from indexing interface).
+
+*. Need to determine if we try to add an API that will support sync primitives.
+    - This is not needed for the initial version of the API.
+    - Requires an indexed integer or datetime field to be updatable, and filterable
+"""
+import abc
+from abc import ABC
+from typing import (
+    Any,
+    Dict,
+    Generic,
+    Iterable,
+    List,
+    NotRequired,
+    Optional,
+    Sequence,
+    TypedDict,
+    TypeVar,
+    Union,
+)
+
+from langchain_core.load.serializable import Serializable
+from langchain_core.runnables import RunnableConfig
+
+
+class IndexableContent(
+    ABC, Serializable
+):  # Introduce new base class for content. Name OK?
+    id: str
```

Comment:
a minor question - what if id doesn't exist and it's created on the vectorstore side?

Reply:
Check out this PR for `BaseMedia`

Ended up merging an Optional ID to allow vectorstores to assign uuids for now. We may make it required in the future so vector stores don't have to implement extra conditional logic on the ID field

Reply:
It seems like it would be better for this to be part of the `BaseMedia`, so that the ID generation is consistent. Specifically, while it would be easy to use a random ID or timestamp, it may be better to hash the content and metadata. That would ensure the ID is consistent and usable for de-duplicating content being inserted multiple times.

Reply:
Yes, this will be the `BaseMedia`.

I am not sure on how to handle the hashes yet. `ids` can be meaningful (e.g., a natural key or a foreign key into another database), so I don't know if we should prevent users from using their own IDs. 

Hashes are obviously useful for de-duplication logic. It's used right now internally in the indexing API: 

https://github.com/langchain-ai/langchain/blob/91b37b2d812eea7fbd2e5c8338c0776826a54c8b/libs/core/langchain_core/indexing/api.py#L56-L56